feat(docs): add some examples of MoveAuthenticator tampering attacks (#10048)
iotaledger/iota: docs/content/developer/move/how-tos/account-abstraction/security/inputs-cryptographic-protection.mdx

---
description: "How to properly protect authenticator input parameters with cryptographic verification to prevent input tampering attacks"
tags:
  - how-to
  - move-sc
  - move-vm
  - account-abstraction
---

# Cryptographic Protection for Authenticator Inputs

import AccountAbstraction from '../../../../../_snippets/developer/move/aa.mdx';

<AccountAbstraction /> 

This how-to explains the importance of cryptographically protecting all authenticator input parameters, not just the transaction digest, to prevent attackers from tampering with inputs after authentication.

## The Vulnerability

> When implementing custom authenticator functions that accept additional input parameters (beyond just verifying the transaction), it's critical to include **all inputs** in the cryptographic verification. If an authenticator only verifies the transaction digest but doesn't cryptographically bind other input parameters (such as shared objects or raw values) to the signature, an attacker can swap those inputs after the signature was created.

Reviewer: a reader could wonder how transaction digest is generated, so maybe it could be useful pointing to somewhere in the docs where transaction digest is explained?

Reviewer 2: [3c6c4ae](https://github.com/iotaledger/iota/pull/10048/commits/3c6c4aeef9e75d87d7e4f49c7fd238edd5c06189)

## Attack Scenario

> Consider an authenticator that accepts a blacklist shared object and a raw value parameter for additional verification logic. If the signature only covers the transaction digest, an attacker can:

Reviewer: ```suggestion
Consider an authenticator that uses a shared `Blacklist` object to reject transactions from blocked accounts. The `Blacklist` stores account addresses that are not allowed to authenticate transactions, and the authenticator checks the current account against that object before allowing the transaction to proceed.

If the `Blacklist` object is passed as an authenticator input, the authenticator must verify that it is the intended blacklist. Otherwise, if the signature only covers the transaction digest, an attacker can:
```

2. Sign the transaction with the authenticator (signature covers transaction digest)
3. **Swap the blacklist shared object** with a different one (e.g., an empty blacklist)
4. **Or change the raw value parameter** to a different value
5. Submit the modified transaction, which will pass signature verification because the transaction digest hasn't changed

## Secure Implementation

Here's an example of a **secure** authenticator that properly protects its inputs:

```move file=<rootDir>/docs/examples/move/abstract_account/authenticator-inputs-cryptographic-protection/sources/account.move#L41-L67
```

**Why this is secure**: The authenticator computes a message hash that includes:
- The transaction digest (`ctx.digest()`)
- The blacklist object ID (`object::id(blacklist)`)  
- The raw value parameter

The signature verification cryptographically binds all these inputs together:

```move file=<rootDir>/docs/examples/move/abstract_account/authenticator-inputs-cryptographic-protection/sources/account.move#L79-L88
```

## Exploitation Example (Why Protection is Necessary)

Without proper input protection, an attacker could swap the blacklist:

> ```rust file=<rootDir>/docs/examples/rust/abstract_account/authenticator-inputs-cryptographic-protection.rs#L413-L447

Reviewer: ```suggestion
```rust file=<rootDir>/docs/examples/rust/abstract_account/authenticator-inputs-cryptographic-protection.rs#L410-L444
```

```

Or swap the raw value parameter:

> ```rust file=<rootDir>/docs/examples/rust/abstract_account/authenticator-inputs-cryptographic-protection.rs#L449-L479

Reviewer: ```suggestion
```rust file=<rootDir>/docs/examples/rust/abstract_account/authenticator-inputs-cryptographic-protection.rs#L446-L476
```

    
    // 1. Include transaction digest
    message.append(*ctx.digest());
    
    // 2. Include shared object ID
    message.append(object::id(blacklist).id_to_bytes());
    
    // 3. Include raw value
    message.append(bcs::to_bytes(&raw_value));
    
    // Hash the combined message
    hash::sha2_256(message)
}
```

### Step 3: Verify Signature Against Complete Message

Use the computed message in signature verification:

```move
assert!(
    ed25519::ed25519_verify(
        &decode(signature),
        account.public_key(),
        &compute_message(blacklist, raw_value, ctx),  // Verify against ALL the inputs
    ),
    EEd25519VerificationFailed,
);
```

### Step 4: Client-Side Signature Creation

When creating transactions on the client side, sign the same message:

```rust
// Prepare the message with all inputs
let mut message = vec![];
message.extend_from_slice(tx_digest.as_ref());
message.extend_from_slice(&blacklist_ref.0.into_bytes());
message.extend_from_slice(bcs::to_bytes(&raw_value)?.as_slice());
let message_hash = Sha256::digest(message.as_slice()).digest;

// Sign the complete message
let signature = keystore.sign_hashed(&signer, &message_hash)?;
```

> ## Key Takeaways

Reviewer 2: I think Key Takeaways could be merged in the best practices (and common pitfalls). This way would reduce duplication and increases the focus of readers in fewer points

Author: fixed in cd9d3eebf7
